source/developer_manual/design/processes.rst

```diff
@@ -0,0 +1,146 @@
+
+-  Master process (dovecot)
+-  Login processes (imap-login, pop3-login)
+-  Authentication process (dovecot-auth)
+-  Mail processes (imap, pop3)
+
+
+Master process
+--------------
+
+This process keeps all the other processes running. If a child process
+dies, another one is restarted automatically. It always runs as root,
+unless you're specifically running everything under a single normal UID.
+
+The master process reads the configuration file and exports the settings
+to other processes via environment variables.
+
+All logging also goes through master process. This avoids problems with
+rotating log files, as there's only a single process to send a signal to
+reopen the log file. Also writing to the same log file (if not using
+syslog) isn't necessarily safe to do in multiple processes concurrently.
+
+Making the logging go through master process also gives a couple of
+advantages from security and reliability point of view: All log lines
+can be prefixed with the process's name and the username of the user who
+was logged in, without the possibility for the process itself to forge
+them. Flooding logs can also be prevented. By default Dovecot allows
+non-privileged processes to write 10 lines per second before it begins
+to delay reading their input, which finally causes the badly behaving
+process to start blocking on writing to stderr instead of eating all the
+CPU and disk space.
+
+In the Dovecot 2.0 design, the master process is split to three parts:
+the Master process which does nothing more than keep the processes
+running, the config process which handles reading the configuration file
+(supporting also eg. SQL storages!) and the log process which handles
+the logging.
+
+
+Login processes
+---------------
+
+The login processes implement the required minimum of the IMAP and POP3
+protocols before a user logs in successfully. There are separate
+processes (and binaries) to handle IMAP and POP3 protocols.
+
+These processes are run with least possible privileges. Unfortunately
+the default UNIX security model still allows them to do much more than
+they would have to: Accept new connections on a socket, connect to new
+UNIX sockets and read and write to existing file descriptors. Still, the
+login process is by default run under a user account that has no special
+access to anything, and inside a non-writable chroot where only a couple
+of files exist. Doing any damage inside there should be difficult.
+
+When a new connection comes, one of the login processes accept()s it.
+After that the client typically does nothing more than ask the server's
+capability list and then log in. The client may also start TLS session
+before logging in.
+
+Authentication is done by talking to the authentication process. The
+login process is completely untrusted by the authentication process, so
+even if an attacker is able to execute arbitrary code inside a login
+process, they won't be able to log in without a valid username and
+password.
```

Comment:
The EXTERNAL mechanism is currently still a notable exception to this. The auth process trusts the login service to determine whether client SSL certificate is valid and what the derived username is.